go-controller/pkg/node/controllers/egressip/egressip.go

```diff
@@ -0,0 +1,1128 @@
+	assignedAddr := sets.New[string]()
+	assignedAddrStrToAddrs := make(map[string]netlink.Addr)
+	assignedIPRoutes := sets.New[string]()
+	assignedIPRouteStrToRoutes := make(map[string]routemanager.RoutesPerLink)
+	assignedIPRules := sets.New[string]()
+	assignedIPRulesStrToRules := make(map[string]netlink.Rule)
+	assignedIPTableV4Rules := sets.New[string]()
+	assignedIPTableV6Rules := sets.New[string]()
+	assignedIPTablesV4StrToRules := make(map[string]iptables.RuleArg)
+	assignedIPTablesV6StrToRules := make(map[string]iptables.RuleArg)
+
+	for _, link := range links {
+		link := link
+		linkName := link.Attrs().Name
+		linkIdx := link.Attrs().Index
+		addresses, err := linkmanager.GetExternallyAvailableAddresses(link, c.v4, c.v6)
+		if err != nil {
+			return fmt.Errorf("unable to get link addresses for link %s: %v", linkName, err)
+		}
+		var assignedAddrFound bool
+		for _, address := range addresses {
+			if address.Label == linkmanager.GetAssignedAddressLabel(linkName) {
+				assignedAddrFound = true
+				addressStr := address.String()
+				assignedAddr.Insert(addressStr)
+				assignedAddrStrToAddrs[addressStr] = address
+			}
+		}
+		if !assignedAddrFound {
+			continue
+		}
+		filter, mask := filterRouteByLinkTable(linkIdx, getRouteTableID(linkIdx))
+		existingRoutes, err := util.GetNetLinkOps().RouteListFiltered(netlink.FAMILY_ALL, filter, mask)
+		if err != nil {
+			return fmt.Errorf("unable to get route list using filter (%s): %v", filter.String(), err)
+		}
+		for _, existingRoute := range existingRoutes {
+			route := routemanager.ConvertNetlinkRouteToRoute(existingRoute)
+			routeStr := route.String()
+			assignedIPRoutes.Insert(routeStr)
+			assignedIPRouteStrToRoutes[routeStr] = routemanager.RoutesPerLink{Link: link,
+				Routes: []routemanager.Route{route}}
+		}
+	}
+	filter, mask := filterRuleByPriority(rulePriority)
+	existingRules, err := util.GetNetLinkOps().RuleListFiltered(netlink.FAMILY_ALL, filter, mask)
+	if err != nil {
+		return fmt.Errorf("failed to list IP rules: %v", err)
+	}
+	for _, existingRule := range existingRules {
+		ruleStr := existingRule.String()
+		assignedIPRules.Insert(ruleStr)
+		assignedIPRulesStrToRules[ruleStr] = existingRule
+	}
+	if c.v4 {
+		ipTableV4Rules, err := c.iptablesManager.GetIPv4ChainRuleArgs(utiliptables.TableNAT, chainName)
+		if err != nil {
+			return fmt.Errorf("failed to list IPTable IPv4 rules: %v", err)
+		}
+		for _, rule := range ipTableV4Rules {
+			ruleStr := strings.Join(rule.Args, " ")
+			assignedIPTableV4Rules.Insert(ruleStr)
+			assignedIPTablesV4StrToRules[ruleStr] = rule
+		}
+	}
+	if c.v6 {
+		ipTableV6Rules, err := c.iptablesManager.GetIPv6ChainRuleArgs(utiliptables.TableNAT, chainName)
+		if err != nil {
+			return fmt.Errorf("failed to list IPTable IPv4 rules: %v", err)
+		}
+		for _, rule := range ipTableV6Rules {
+			ruleStr := strings.Join(rule.Args, " ")
+			assignedIPTableV6Rules.Insert(ruleStr)
+			assignedIPTablesV6StrToRules[ruleStr] = rule
+		}
+	}
```

Comment:
Should we not check for ip family config mode and gather ip table rules for both families to clean up the ones that might be stale due to a dual-stack -> single stack config change?